REGRESSION(291435@main...291445@main): [iOS Debug] TestWebKitAPI.IndexedDB.IndexedDBSuspendImminently is a failure/timeout (flaky in EWS)

achristensen07 opened the pull request:
#### a899e6f6ed50d213260227bfac861f67e5b60e54
<pre>
REGRESSION(291435@main...291445@main): [iOS Debug] TestWebKitAPI.IndexedDB.IndexedDBSuspendImminently is a failure/timeout (flaky in EWS)
<a href="https://bugs.webkit.org/show_bug.cgi?id=290203">https://bugs.webkit.org/show_bug.cgi?id=290203</a>
<a href="https://rdar.apple.com/147599411">rdar://147599411</a>

Reviewed by Sihui Liu.

TestWebKitAPI::Util::run doesn&apos;t necessarily stop everything when the condition becomes true.
There is a chance that more work can be done before it returns.
When multiple messages are awaited, sometimes we&apos;ll wait for the first one and the first
and second one are received before TestWebKitAPI::Util::run returns, then when we try to wait
for the second one, no more messages come.  I introduce TestScriptMessageHandler with a Deque
to make it more robust.  If it has already received a message, it just returns it immediately.
If it receives too many messages, it&apos;ll hang on to them until later.
A similar techinque is already used in TestWKWebView.waitForMessages

* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestScriptMessageHandler.mm: Added.
(-[TestScriptMessageHandler waitForMessage]):
(-[TestScriptMessageHandler userContentController:didReceiveScriptMessage:]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IndexedDBSuspendImminently.mm:
(TEST(IndexedDB, IndexedDBSuspendImminently)):
(TEST(IndexedDB, SuspendImminentlyForThirdPartyDatabases)):
(-[IndexedDBSuspendImminentlyMessageHandler userContentController:didReceiveScriptMessage:]): Deleted.
(runUntilMessageReceived): Deleted.
(keepNetworkProcessActive): Deleted.
(TEST(IndexedDB, DISABLED_IndexedDBSuspendImminently)): Deleted.
* Tools/TestWebKitAPI/cocoa/TestScriptMessageHandler.h: Added.
* Tools/TestWebKitAPI/cocoa/TestScriptMessageHandler.mm: Added.
(-[TestScriptMessageHandler waitForMessage]):
(-[TestScriptMessageHandler userContentController:didReceiveScriptMessage:]):

Canonical link: <a href="https://commits.webkit.org/292658@main">https://commits.webkit.org/292658@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c313a331c177fe3acb1eaea67cd55c440234a6a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96621 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16235 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6342 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101693 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47141 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98666 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16531 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24669 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73629 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30853 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99624 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12438 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87370 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53965 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12197 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5163 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46469 "Built successfully") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82298 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5256 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103717 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23688 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17264 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82680 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23939 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83419 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82055 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26713 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4239 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17162 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15578 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23651 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23310 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26790 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25051 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/IndexedDBSuspendImminently.mm:
```diff
@@ -39,60 +40,27 @@
-
-- (void)userContentController:(WKUserContentController *)userContentController didReceiveScriptMessage:(WKScriptMessage *)message
-{
-    idbActivitiesStarted = true;
-    receivedScriptMessage = true;
-    lastScriptMessage = message;
```
Can we just do here: 
```
receivedScriptMessage = true;
scriptMessages.append(message);
```
And use `[getNextMessage() body]` below?
(There is a `getNextMessage` in DeprecatedGlobalValues.h)

achristensen07 replied:
We could, but this is a non-global replacement for DeprecatedGlobalValues.  I think it would be better to replace all use of getNextMessage() with TestScriptMessageHandler

achristensen07 committed:
REGRESSION(291435@main...291445@main): [iOS Debug] TestWebKitAPI.IndexedDB.IndexedDBSuspendImminently is a failure/timeout (flaky in EWS)
https://bugs.webkit.org/show_bug.cgi?id=290203
rdar://147599411

Reviewed by Sihui Liu.

TestWebKitAPI::Util::run doesn't necessarily stop everything when the condition becomes true.
There is a chance that more work can be done before it returns.
When multiple messages are awaited, sometimes we'll wait for the first one and the first
and second one are received before TestWebKitAPI::Util::run returns, then when we try to wait
for the second one, no more messages come.  I introduce TestScriptMessageHandler with a Deque
to make it more robust.  If it has already received a message, it just returns it immediately.
If it receives too many messages, it'll hang on to them until later.
A similar techinque is already used in TestWKWebView.waitForMessages

* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestScriptMessageHandler.mm: Added.
(-[TestScriptMessageHandler waitForMessage]):
(-[TestScriptMessageHandler userContentController:didReceiveScriptMessage:]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IndexedDBSuspendImminently.mm:
(TEST(IndexedDB, IndexedDBSuspendImminently)):
(TEST(IndexedDB, SuspendImminentlyForThirdPartyDatabases)):
(-[IndexedDBSuspendImminentlyMessageHandler userContentController:didReceiveScriptMessage:]): Deleted.
(runUntilMessageReceived): Deleted.
(keepNetworkProcessActive): Deleted.
(TEST(IndexedDB, DISABLED_IndexedDBSuspendImminently)): Deleted.
* Tools/TestWebKitAPI/cocoa/TestScriptMessageHandler.h: Added.
* Tools/TestWebKitAPI/cocoa/TestScriptMessageHandler.mm: Added.
(-[TestScriptMessageHandler waitForMessage]):
(-[TestScriptMessageHandler userContentController:didReceiveScriptMessage:]):

Canonical link: https://commits.webkit.org/292658@main
